docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,309 @@
+    - [Option 3: CAPZ resource defines an entire unstructured ASO resource inline](#option-3-capz-resource-defines-an-entire-unstructured-aso-resource-inline)
+    - [Option 4: CAPZ resource defines an entire typed ASO resource inline](#option-4-capz-resource-defines-an-entire-typed-aso-resource-inline)
+    - [Decision](#decision)
+  - [Security Model](#security-model)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Alternatives](#alternatives)
+- [Upgrade Strategy](#upgrade-strategy)
+- [Additional Details](#additional-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Glossary
+
+## Summary
+
+CAPZ's AzureManagedControlPlane and AzureManagedMachinePool resources expose AKS's managed cluster and agent
+pool resources for Cluster API. Currently, new features in AKS require manual changes to teach CAPZ about
+those features for users to be able to take advantage of them natively in Cluster API. As a result, there are
+several AKS features available that cannot be used from Cluster API. This proposal describes how CAPZ will
+automatically make all AKS features available on an ongoing basis with minimal maintenance.
+
+## Motivation
+
+Historically, CAPZ has exposed an opinionated subset of AKS features that are tested and known to work within
+the Cluster API ecosystem. Since then, it has become increasingly clear that new AKS features are generally
+suitable to implement in CAPZ and users are interested in having all AKS features available to them from
+Cluster API.
+
+When gaps exist in the set of features available in AKS and what CAPZ offers, users of other infrastructure
+management solutions may not be able to adopt CAPZ. If all AKS features could be used from CAPZ, this would
+not be an issue.
+
+The AKS feature set changes rapidly alongside CAPZ's users' desire to utilize those features. Because making
+new AKS features available in CAPZ requires a considerable amount of mechanical, manual effort to implement,
+review, and test, requests for new AKS features account for a large portion of the cost to maintain CAPZ.
+
+Another long-standing feature request in CAPZ is the ability to adopt existing AKS clusters into management by
+Cluster API (https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1173). Making the entire AKS
+API surface area available from CAPZ is required to enable this so existing clusters' full representations can
+be reflected in Cluster API.
+
+### Goals
+
+- Narrow the gap between the sets of features offered by AKS and CAPZ.
+- Reduce the maintenance cost of making new AKS features available from CAPZ.
+
+### Non-Goals/Future Work
+
+- Automate the features available from other Azure services besides AKS.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
```

Comment:
Also thinking of another user story where we should decide if is in or out of scope:
As a CAPZ managed cluster user, I'd like to keep my existing CAPZ cluster definition and just add the missing features I need until those features are made available in CAPZ definitions natively.

There are different ways we could word this, but reading through the options - I can see some of them are more "placeholder" references to ASO code for an entire cluster versus a hybrid approach.

Reply:
It should still be possible today in CAPZ 1.12 for a user to create an AKS cluster with CAPZ and then define extra fields (even in a newer API version) that CAPZ doesn't know about and CAPZ will preserve those. The main limitation there currently is in which AKS API versions are available with ASO. Does that cover what you mean here?

Reply:
If there is a pseudo-supported path today where you have a cluster defined in CAPZ YAML AND the same cluster defined in ASO YAML from the same management cluster, but the ASO YAML only has extra fields... That definitely should be called out in this document - perhaps in the "alternative" section.

Reply:
The main reason we can't rely on that as a long-term solution is because CAPZ still creates the ASO ManagedCluster at runtime (some time after an AzureManagedControlPlane is created), so modifying the ASO resource after the fact wouldn't enable a single `kubectl apply` to fully specify an AKS cluster if the CAPZ API is still limited to only a subset of fields. It also doesn't work at all to set parameters that must be specified at cluster creation time and are immutable after that.

Going back to your original point, I think it is worthwhile to make the solution here backwards compatible and can be adopted iteratively, so I clarified that in the goals.